### ISSUE TITLE
Client: Batching

### ISSUE THREAD

**Issue by sentientwaffle** [tigerbeetle/tigerbeetle#489]
### Batching Background

Batching requests is essential for good throughput & latency.

Right now applications must batch accounts/transfers/queries themselves before invoking the client. The client ships that whole batch to the cluster in a single message.

The goal of this design is to build transparent batching into all client implementations so that users don't need to think about them so much. (At least until they need to eke out more performance).

### Client Background

Currently the client architecture is:

  1. Language client, which is built atop the
  2. C client, which is built atop the
  3. Zig client (`vsr.Client`).

`vsr.Client`'s responsibilities are currently:

  - Sending requests, receiving replies.
  - Retrying requests
  - Automatic session registration
  - Handle session eviction
  - Request queueing
  - View discovery (ping/pong)

The C client's responsibilities:

  - Concurrency for language runtimes.

The language client's responsibilities:

  - Encode requests into bytes.
  - Decode replies from bytes.

## Terminology

- _logical batch_: A sequence of one or more events which will be sent together in a single request message.
    - Logical batches are created by the application.
    - The client is not allowed to split a logical batch across multiple request messages.
    - All of the events in a logical batch have a common `Operation`.
- _physical batch_: A sequence of one or more logical batches which are grouped together in a single request message.
    - Physical batches are created by the client itself.
    - All of the logical batches in a physical batch have a common `Operation`.

## Design

`vsr.Client` is additionally responsible for batching.

Currently,

```
const message: *Message = client.get_message();
... write events to message.body()
client.request(user_data, callback, operation, message, message_body_size)
```

Replace with:

```
const batch: BatchLogical(operation) = client.get_batch(operation, number of events);
... write events (the logical batch) to batch.slice() (a []u8.. or maybe it could be a []StateMachine.Event(operation) if that works)
... The memory returned by `Batch.slice()` points to a slice of message body memory.
Client.batch(user_data, callback, batch)
```

When `.get_batch()` is called:

1. Find the first `Batch` (aka batch) in the queue with a matching `operation` that has room for the entire `batch_logical`.
2. Copy the `batch_logical`'s bytes into that memory (or deserialize the accounts/transfers into that memory, depending on the language).
3. Save the `callback`+ `user_data` along with the range of events so that they can be used to demux the reply.
4. (If there is no such batch/message, create a new one and append it to the queue).

In `.batch()`:
1. Any time there is no request in flight, send the first batch immediately.

In the client `tick`:
1. Any time there is no request in flight, send the first request in the queue.
2. When the `Client` gets a reply back, it demux the reply up into its component replies and fan them out to the different logical batches' callbacks.

Notes:
- `.request()` is now `.batch()` since invoking it is not 1:1 with request messages.
- The expectation is that the `batch` is passed back into `Client.batch()` pretty much immediately, and then never used again.
- (The "always keep 1 request in flight" rule is similar to Nagle's algorithm).

##### Batch size

"How many events to put in a request message" isn't simply "how many can fit" — for example, the query API (not yet implemented) will not use batching in requests. This logic/decision is part of the state machine interface (currently `StateMachine.constants.batch_max.{operation}`).

##### Reply encoding/decoding

Decoding replies (and requests) is statemachine/operation-specific (e.g. the `.{.index = 123, .code = .reserved_flag}` format for the accounting state machine's `operation=create_transfers` reply).

The `StateMachine` API needs to expose a function, which the `Client` can use to split the reply up into results and map them back to the original logical batches.

## Example

(Examples written for the Node.js client.)

Assuming the following is called while a request is already in flight,
this would send a single request message, and receive a single reply message:

```javascript
client.create_transfer([A,B,C]).then(...)
client.create_transfer([D,E,F]).then(...)
```

(Note that lack of an `await` — if there was an `await`, then of course the former would finish before the latter began, so they could not be batched together.)

Likewise,

```javascript
client.create_transfer([A,B,C]).then(...)
client.create_accounts([Q,R,S]).then(...)
client.create_transfer([D,E,F]).then(...)
client.create_transfer([T,U,V]).then(...)
```

results in two messages:

1. `create_transfers [A,B,C,D,E,F]`
2. `create_accounts [Q,R,S,T,U,V]`

**Comment by jorangreef** [comment]
> In the Client tick: Any time there is no request in flight, send the first request in the queue.

If I understand correctly, this would introduce a little bit of lag between calling `Client.batch()` and when `Client.tick()` is called.

I think we should rather fire off the request immediately, if there's nothing inflight, rather than waiting for `tick()` to do so.

**Comment by sentientwaffle** [comment]
> I think we should rather fire off the request immediately, if there's nothing inflight, rather than waiting for tick() to do so.

:+1: good catch, I'll update the doc.

**Comment by batiati** [comment]
From the text above:

>When .get_batch() is called:
>
>Find the first Batch (aka batch) in the queue with a matching operation that has room for the entire batch_logical.

Consider this scenario:

```javascript
client.create_accounts([A, B]).then(...) // Creates 2 accounts A and B
client.create_transfer([A2B]).then(...)  // Transfers from account A to B
client.create_accounts([C, D]).then(...) // Creates 2 accounts C and D
client.create_transfer([C2D]).then(...) // Transfers from C to D
```
results in the following messages:

1. Accounts `A` and `B` will be sent immediately since there is no request in flight.
2. Transfer `A2B` will be enqueued.
3. Accounts `C` and `D` will be enqueued in another physical batch since the first in the queue has a different operation.
4. Transfer `C2D` will be batched in the same physical batch as `A2B` (step 2), since it is the first in the queue with a matching operation.

The transfer `C2D` was reordered and will be sent before the accounts `C` and `D`, resulting in failure due to accounts not being found.

I think we can only safely batch the last physical batch in the queue.

**Comment by jamii** [comment]
This is how most async apis behave even without batching though eg if every call to client.create_* was a separate http request. I think it's pretty well engrained that if you care about ordering you have to write something like:

```js
client.create_accounts([C, D]).then(() => {
  client.create_transfer([C2D]).then(...);
})
```

Or equivalently:

```js
await client.create_accounts([C, D])
await client.create_transfer([C2D])
```

Where batching becomes interesting is when you have many independent tasks doing something like the above. The relative ordering between them is undefined anyway, so we're free to choose orderings that give us better batching. (Probably within limits of fairness/anti-starvation of course).

**Comment by batiati** [comment]
> This is how most async apis behave even without batching though eg if every call to client.create_* was a separate http request. I think it's pretty well engrained that if you care about ordering you have to write something like:

I was trying a find any situation where batching could change the request's semantics, but today morning I woke up with this feeling: **"It would be an application misuse and not a batching flaw"**.

It's refreshing to see your reply here confirming it 😅🙌

**Comment by jorangreef** [comment]
Great to see you coming up with the hacker's counter-example... and then sleeping well on it! 😅 🖖

**Comment by batiati** [comment]
Closed by #1257
